File: apps/android/app/src/main/kotlin/sh/nikhil/swekitty/SessionNaming.kt
package sh.nikhil.swekitty

import java.time.Instant
import java.time.LocalDate
import java.time.ZoneId
import java.time.format.DateTimeFormatter
import java.util.Locale
import uniffi.conduit_core.ConversationItem
import uniffi.conduit_core.ProjectSession

/**
 * Pure, Compose-free helpers for turning a raw session into a friendly,
 * scannable label — and for bucketing a history list by recency. Lifted
 * out of [SessionStore] / the screens so JUnit can pin the rules with
 * fixed clocks (no Robolectric, no live ViewModel). Android parity of the
 * iOS list/naming work; mirrors the priority order documented on
 * `SessionStore.displayName`.
 */
internal object SessionNaming {

    /** Max characters for a chat-derived name before ellipsizing. */
    const val NAME_CHAR_LIMIT = 40

    /**
     * Friendly default name for a session, in priority order:
     *  1. `custom` (a user-set rename) — wins, returned verbatim.
     *  2. `aiTitle` (the broker AI-generated title, task: ai-session-titles)
     *     when it isn't itself the raw UUID.
     *  3. First user chat message — trimmed, single-lined, ellipsized.
     *  4. `serverLabel` (a broker-supplied `displayName`/`sessionName`),
     *     when it isn't itself the raw UUID.
     *  5. Fallback: "<agent> · <relative start time>" from `startedAt`.
     *
     * NEVER returns the raw session id / UUID `name`. The UUID is only for
     * Session Info, never a user-facing label.
     */
    fun friendly(
        sessionId: String,
        rawName: String,
        agent: String,
        custom: String?,
        firstUserMessage: String?,
        serverLabel: String?,
        startedAt: String?,
        aiTitle: String? = null,
        nowMs: Long = System.currentTimeMillis(),
        zone: ZoneId = ZoneId.systemDefault(),
    ): String {
        custom?.trim()?.takeIf { it.isNotEmpty() }?.let { return it }

        // Broker AI title sits just below a manual rename and above the
        // first message. Reject a UUID-shaped echo so we never re-spill the
        // id as a label.
        aiTitle?.trim()
            ?.takeIf { it.isNotEmpty() && !looksLikeRawId(it, sessionId, rawName) }
            ?.let { return it }

        firstUserMessage?.let { msg ->
            condense(msg)?.let { return it }
        }

        // A broker-supplied label is only useful when it isn't the UUID we
        // were trying to avoid (the broker echoes the id as `name` for
        // unnamed sessions).
        serverLabel?.trim()
            ?.takeIf { it.isNotEmpty() && !looksLikeRawId(it, sessionId, rawName) }
            ?.let { return it }

        return fallbackName(agent, startedAt, nowMs, zone)
    }

    /**
     * Trim a chat message down to a single ellipsized line. Collapses
     * internal whitespace/newlines to single spaces (a multi-line prompt
     * should read as one tidy line), returns null when the message is
     * effectively empty so the caller can fall through.
     */
    fun condense(raw: String): String? {
        val oneLine = raw.replace(Regex("\\s+"), " ").trim()
        if (oneLine.isEmpty()) return null
        if (oneLine.length <= NAME_CHAR_LIMIT) return oneLine
        // Trim to the limit, then drop a dangling partial word for cleaner
        // truncation before appending the ellipsis.
        val clipped = oneLine.take(NAME_CHAR_LIMIT).trimEnd()
        val lastSpace = clipped.lastIndexOf(' ')
        val base = if (lastSpace >= NAME_CHAR_LIMIT / 2) clipped.substring(0, lastSpace) else clipped
        return base.trimEnd() + "…"
    }

    /** "<agent> · <relative start time>", e.g. "claude · 4:02 PM" / "claude · Mon". */
    fun fallbackName(
        agent: String,
        startedAt: String?,
        nowMs: Long = System.currentTimeMillis(),
        zone: ZoneId = ZoneId.systemDefault(),
    ): String {
        val agentLabel = agent.trim().ifEmpty { "session" }
        val when_ = relativeStart(startedAt, nowMs, zone)
        return if (when_ != null) "$agentLabel · $when_" else agentLabel
    }

    /**
     * Clock-style relative start: time of day when the session started
     * today ("4:02 PM"), the weekday for anything older ("Mon"), or a
     * short date when older than a week ("May 12"). Null when unparseable.
     */
    fun relativeStart(
        raw: String?,
        nowMs: Long = System.currentTimeMillis(),
        zone: ZoneId = ZoneId.systemDefault(),
    ): String? {
        val instant = parseInstant(raw) ?: return null
        val started = instant.atZone(zone).toLocalDate()
        val today = Instant.ofEpochMilli(nowMs).atZone(zone).toLocalDate()
        val days = today.toEpochDay() - started.toEpochDay()
        val dt = instant.atZone(zone)
        return when {
            days <= 0L -> dt.format(TIME_FMT)
            days < 7L -> dt.format(WEEKDAY_FMT)
            else -> dt.format(SHORT_DATE_FMT)
        }
    }

    /**
     * Short "time ago" string for a row's trailing slot ("now", "2m",
     * "3h", "5d", or a short date). Empty when unparseable/blank.
     */
    fun relativeAgo(
        raw: String?,
        nowMs: Long = System.currentTimeMillis(),
        zone: ZoneId = ZoneId.systemDefault(),
    ): String {
        val instant = parseInstant(raw) ?: return ""
        val deltaMs = (nowMs - instant.toEpochMilli()).coerceAtLeast(0)
        val minutes = deltaMs / 60_000
        val hours = deltaMs / 3_600_000
        val days = deltaMs / 86_400_000
        return when {
            minutes < 1 -> "now"
            minutes < 60 -> "${minutes}m ago"
            hours < 24 -> "${hours}h ago"
            days < 7 -> "${days}d ago"
            else -> instant.atZone(zone).format(SHORT_DATE_FMT)
        }
    }

    private fun looksLikeRawId(label: String, sessionId: String, rawName: String): Boolean {
        if (label == sessionId || label == rawName) return true
        // A bare UUID: 8-4-4-4-12 hex. Such a label is never something a
        // human typed, so treat it as "raw" and fall through.
        return UUID_REGEX.matches(label)
    }

    private fun parseInstant(raw: String?): Instant? {
        val trimmed = raw?.trim().orEmpty()
        if (trimmed.isEmpty()) return null
        return runCatching { Instant.parse(trimmed) }.getOrNull()
            ?: runCatching {
                java.time.OffsetDateTime.parse(trimmed).toInstant()
            }.getOrNull()
    }

    /** Convenience: the friendly name for a live [ProjectSession]. */
    fun friendlyFor(
        session: ProjectSession,
        custom: String?,
        firstUserMessage: String?,
        aiTitle: String? = null,
        nowMs: Long = System.currentTimeMillis(),
    ): String = friendly(
        sessionId = session.id,
        rawName = session.name,
        agent = session.assistant,
        custom = custom,
        firstUserMessage = firstUserMessage,
        serverLabel = session.displayName,
        startedAt = session.startedAt,
        aiTitle = aiTitle,
        nowMs = nowMs,
    )

    val UUID_REGEX = Regex(
        "[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12}",
    )
    private val TIME_FMT = DateTimeFormatter.ofPattern("h:mm a", Locale.US)
    private val WEEKDAY_FMT = DateTimeFormatter.ofPattern("EEE", Locale.US)
    private val SHORT_DATE_FMT = DateTimeFormatter.ofPattern("MMM d", Locale.US)

    /**
     * Condense a single transcript item into a one-line activity preview
     * for the home card. Pure string inputs (role / kind / toolName /
     * command / content) keep this Compose- and core-type-free so JUnit
     * can pin it. Mirror of iOS `HomeViewModel.activityPreview`:
     *  - a tool action surfaces its command ("Bash: cargo test") or, with
     *    no command, "<tool>: <first line>";
     *  - assistant / other text uses the first non-empty body line.
     * Returns null when there's nothing worth showing.
     */
    fun activityPreview(
        role: String,
        kind: String,
        toolName: String?,
        command: String?,
        content: String,
        budget: Int = 72,
    ): String? {
        if (role.lowercase() == "tool" || kind.lowercase() == "tool") {
            val cmd = command?.takeIf { it.isNotBlank() }
            if (cmd != null) {
                val label = toolName?.takeIf { it.isNotEmpty() } ?: "Run"
                return clip("$label: ${firstLine(cmd)}", budget)
            }
            val body = firstLine(content)
            if (body.isNotEmpty()) {
                val prefix = toolName?.takeIf { it.isNotEmpty() }?.let { "$it: " } ?: ""
                return clip(prefix + body, budget)
            }
            return toolName?.takeIf { it.isNotEmpty() }?.let { clip(it, budget) }
        }
        val body = firstLine(content)
        return if (body.isEmpty()) null else clip(body, budget)
    }

    /**
     * First non-empty line of a (possibly multi-line) string with
     * internal whitespace runs collapsed to single spaces.
     */
    private fun firstLine(raw: String): String {
        raw.lineSequence().forEach { line ->
            val collapsed = line.split(Regex("\\s+")).filter { it.isNotEmpty() }.joinToString(" ")
            if (collapsed.isNotEmpty()) return collapsed
        }
        return ""
    }

    /** Trim to [budget] chars with an ellipsis when over. */
    private fun clip(s: String, budget: Int): String =
        if (s.length <= budget) s else s.take(budget - 1) + "…"
}

/**
 * Recency buckets for the history list. Order is fixed: only non-empty
 * buckets are emitted, in this sequence. Android parity of the iOS
 * list/naming PR's time grouping (which replaced server-grouping).
 */
enum class RecencyBucket(val label: String) {
    TODAY("Today"),
    YESTERDAY("Yesterday"),
    PREVIOUS_7_DAYS("Previous 7 Days"),
    EARLIER("Earlier"),
}

/**
 * Pure grouping of arbitrary rows by a `lastSeen` timestamp into
 * [RecencyBucket]s, latest-first within each bucket. Generic over the row
 * type so both the search screen and tests can drive it; callers supply a
 * `lastSeen` extractor (RFC3339 string) and the bucketing key. Empty
 * buckets are dropped; bucket order follows [RecencyBucket.values].
 */
object SessionRecencyGrouping {

    data class Group<T>(val bucket: RecencyBucket, val rows: List<T>)

    fun <T> group(
        rows: List<T>,
        nowMs: Long = System.currentTimeMillis(),
        zone: ZoneId = ZoneId.systemDefault(),
        lastSeen: (T) -> String?,
    ): List<Group<T>> {
        val today = Instant.ofEpochMilli(nowMs).atZone(zone).toLocalDate()
        // Decorate each row with (bucket, epochMilli) so we can sort
        // latest-first deterministically. Rows with no parseable timestamp
        // sort to the bottom of EARLIER.
        val decorated = rows.map { row ->
            val instant = parseInstant(lastSeen(row))
            val bucket = bucketFor(instant, today, zone)
            Triple(row, bucket, instant?.toEpochMilli() ?: Long.MIN_VALUE)
        }
        return RecencyBucket.entries.mapNotNull { bucket ->
            val inBucket = decorated
                .filter { it.second == bucket }
                .sortedByDescending { it.third }
                .map { it.first }
            if (inBucket.isEmpty()) null else Group(bucket, inBucket)
        }
    }

    private fun bucketFor(instant: Instant?, today: LocalDate, zone: ZoneId): RecencyBucket {
        if (instant == null) return RecencyBucket.EARLIER
        val date = instant.atZone(zone).toLocalDate()
        val days = today.toEpochDay() - date.toEpochDay()
        return when {
            days <= 0L -> RecencyBucket.TODAY
            days == 1L -> RecencyBucket.YESTERDAY
            days < 8L -> RecencyBucket.PREVIOUS_7_DAYS
            else -> RecencyBucket.EARLIER
        }
    }

    private fun parseInstant(raw: String?): Instant? {
        val trimmed = raw?.trim().orEmpty()
        if (trimmed.isEmpty()) return null
        return runCatching { Instant.parse(trimmed) }.getOrNull()
            ?: runCatching { java.time.OffsetDateTime.parse(trimmed).toInstant() }.getOrNull()
    }
}

/**
 * Best-effort first user message for a session, scanning the typed
 * conversation log then the raw chat log — mirror of iOS
 * `SessionStore.firstUserMessage(in:)`. Returns null when neither carries
 * a user turn yet (terminal-only sessions, or a transcript not loaded).
 */
internal fun firstUserMessageOf(
    conversation: List<ConversationItem>?,
): String? {
    conversation?.firstOrNull { it.role.lowercase() == "user" }?.content
        ?.takeIf { it.isNotBlank() }
        ?.let { return it }
    return null
}

/**
 * One-line "what's happening" preview for a session card — the most
 * recent NON-user transcript item (assistant reply or tool action),
 * condensed. The card title is already the first user message, so this
 * complements it and lets the user tell active sessions apart at a
 * glance. Mirror of iOS `LitterHomeView.latestActivityPreview` +
 * `HomeViewModel.activityPreview`. Returns null when there's no non-user
 * activity yet (placeholder / user-only transcripts).
 */
internal fun latestActivityPreviewOf(
    conversation: List<ConversationItem>?,
): String? {
    val latest = conversation?.lastOrNull { it.role.lowercase() != "user" } ?: return null
    return SessionNaming.activityPreview(
        role = latest.role,
        kind = latest.kind,
        toolName = latest.toolName,
        command = latest.command,
        content = latest.content,
    )
}

File: apps/ios/Sources/Shared/SessionNaming.swift
import Foundation

/// Friendly, user-facing session names. The broker mints a raw UUID as a
/// session's `name` (and often echoes that same UUID back as the broker
/// `sessionName`/`displayName` label), so naively rendering `session.name`
/// — or a broker label that's really the id — spills a UUID into the home
/// list, the history list, and project headers. That's the bug this type
/// fixes.
///
/// The resolution priority (see `SessionStore.displayName(for:)` /
/// `SessionsScreen.rowTitle`) is:
///   1. A genuine user-set custom name (never a UUID) — wins.
///   2. The first user chat message, trimmed to one ellipsized line.
///   3. A fallback of `"<agent> · <relative start time>"`.
///
/// All helpers are pure + `nonisolated static` so they can be unit-tested
/// (`SessionNamingTests`) without a live store, and reused from both the
/// live home list and the persisted history screen.
enum SessionNaming {

    /// Max characters for a chat-message-derived title before we
    /// ellipsize. ChatGPT/Claude-style short single-line label.
    static let titleBudget = 40

    /// True when `text` looks like a bare session id — a canonical
    /// 36-char hyphenated UUID (`8299a0d1-eabe-4801-9a5f-ffea9eec60f7`)
    /// or an exact match for the supplied session id. We treat both the
    /// stored `displayNames[id]` entry AND any broker-supplied label as
    /// suspect because the broker commonly sets the label to the id; a
    /// UUID-shaped string is never something the user typed, so it can
    /// never be a "custom name".
    nonisolated static func looksLikeRawID(_ text: String, sessionID: String) -> Bool {
        let trimmed = text.trimmingCharacters(in: .whitespacesAndNewlines)
        if trimmed.isEmpty { return true }
        if trimmed == sessionID { return true }
        return isUUIDLike(trimmed)
    }

    /// Canonical 8-4-4-4-12 hex UUID shape, case-insensitive.
    nonisolated static func isUUIDLike(_ text: String) -> Bool {
        let parts = text.split(separator: "-", omittingEmptySubsequences: false)
        guard parts.count == 5 else { return false }
        let expectedLengths = [8, 4, 4, 4, 12]
        for (part, expected) in zip(parts, expectedLengths) {
            guard part.count == expected else { return false }
            guard part.allSatisfy({ $0.isHexDigit }) else { return false }
        }
        return true
    }

    /// Collapse a chat message into a single short title line: first
    /// non-empty line, internal whitespace runs collapsed to single
    /// spaces, trimmed, then ellipsized to `titleBudget`. Returns nil for
    /// an empty/whitespace-only message so callers fall through to the
    /// agent+time fallback.
    nonisolated static func titleFromMessage(_ raw: String) -> String? {
        let firstLine = raw
            .split(whereSeparator: { $0.isNewline })
            .first
            .map(String.init) ?? raw
        let collapsed = firstLine
            .split(whereSeparator: { $0.isWhitespace })
            .joined(separator: " ")
        let cleaned = collapsed.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !cleaned.isEmpty else { return nil }
        if cleaned.count <= titleBudget { return cleaned }
        return String(cleaned.prefix(titleBudget - 1)) + "…"
    }

    /// Fallback label for a session with no chat yet (terminal-only, or
    /// the transcript hasn't streamed): `"<agent> · <relative time>"`.
    /// Today → time of day ("claude · 4:02 PM"); within the last week →
    /// weekday ("claude · Mon"); older → short date. `startedAt` is an
    /// RFC3339 string; when missing/unparseable we degrade to just the
    /// agent name (never a UUID).
    nonisolated static func fallbackName(
        agent: String,
        startedAt: String?,
        now: Date = Date(),
        calendar: Calendar = .current
    ) -> String {
        let agentLabel = agent.trimmingCharacters(in: .whitespacesAndNewlines)
        let safeAgent = agentLabel.isEmpty ? "session" : agentLabel
        guard let raw = startedAt,
              let date = parseTimestamp(raw) else {
            return safeAgent
        }
        return "\(safeAgent) · \(relativeStamp(for: date, now: now, calendar: calendar))"
    }

    /// Whole-calendar-day distance from `date` to `now` (0 = same day as
    /// `now`, 1 = the day before `now`, …). Computed against the injected
    /// `now` rather than the device clock so it's deterministic in tests —
    /// `Calendar.isDateInToday`/`isDateInYesterday` ignore any anchor and
    /// always compare to the real `Date()`, which is exactly the bug this
    /// avoids.
    nonisolated static func dayDistance(
        from date: Date,
        to now: Date,
        calendar: Calendar
    ) -> Int? {
        let startNow = calendar.startOfDay(for: now)
        let startDate = calendar.startOfDay(for: date)
        return calendar.dateComponents([.day], from: startDate, to: startNow).day
    }

    /// Human relative stamp used by the fallback name. Deterministic given
    /// an injected `now`/`calendar` so it's unit-testable.
    nonisolated static func relativeStamp(
        for date: Date,
        now: Date = Date(),
        calendar: Calendar = .current
    ) -> String {
        let distance = dayDistance(from: date, to: now, calendar: calendar) ?? Int.max
        if distance <= 0 {
            // Same day as `now` (or future) → time of day.
            let f = formatter(calendar)
            f.timeStyle = .short
            f.dateStyle = .none
            return f.string(from: date)
        }
        if distance == 1 {
            return "Yesterday"
        }
        // Within the last 7 days → weekday name ("Mon").
        if distance < 7 {
            let f = formatter(calendar)
            f.dateFormat = "EEE"
            return f.string(from: date)
        }
        let f = formatter(calendar)
        f.dateStyle = .short
        f.timeStyle = .none
        return f.string(from: date)
    }

    /// A `DateFormatter` that honors the supplied calendar's timezone +
    /// locale so the rendered stamp is deterministic for an injected
    /// calendar (tests) and correct for the device calendar (production).
    private nonisolated static func formatter(_ calendar: Calendar) -> DateFormatter {
        let f = DateFormatter()
        f.calendar = calendar
        f.locale = calendar.locale ?? .current
        f.timeZone = calendar.timeZone
        return f
    }

    /// A meaningful, user-pickable cwd to surface in the UI, or nil for
    /// the ephemeral per-session scratch dir. The broker runs each
    /// session in `…/.conduit/sessions/<id>/work`, which is not a real
    /// project directory — showing it as if it were is the bug. We hide
    /// any path that lives under a `…/sessions/<…>/work` (or bare
    /// `…/sessions/<…>`) segment so only a genuine repo cwd surfaces.
    nonisolated static func meaningfulWorkingDir(_ cwd: String?) -> String? {
        guard let cwd else { return nil }
        let trimmed = cwd.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else { return nil }
        let lower = trimmed.lowercased()
        if lower.contains(".conduit/sessions/") { return nil }
        // Generic guard for the `/sessions/<id>/work` shape regardless of
        // the leading root.
        let normalized = lower.hasSuffix("/") ? String(lower.dropLast()) : lower
        if normalized.hasSuffix("/work"),
           normalized.contains("/sessions/") {
            return nil
        }
        return trimmed
    }

    /// Tolerant RFC3339 parse — accepts the fractional-second variant the
    /// broker sometimes emits as well as the plain form.
    nonisolated static func parseTimestamp(_ raw: String) -> Date? {
        let withFraction = ISO8601DateFormatter()
        withFraction.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        if let d = withFraction.date(from: raw) { return d }
        let plain = ISO8601DateFormatter()
        plain.formatOptions = [.withInternetDateTime]
        return plain.date(from: raw)
    }
}
